File: vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

bool allalpha(string s);

int main(int argc, string argv[])
{
    if (argc != 2 || !allalpha(argv[1]))                // WRONG USAGE
    {
        printf("Usage: ./caesar <alphabetical keyword>\n");
        return 1;
    }

    string keystr = argv[1];                            // RIGHT USAGE
    int keylen = strlen(keystr), keyarr[keylen];
    for (int i = 0; i < keylen; i++)                            // array of keys from the keyword
    {
        if (isupper(keystr[i]))
        {
            keyarr[i] = (int) keystr[i] - 65;
        }

        else
        {
            keyarr[i] = (int) keystr[i] - 97;
        }
    }

    string plainstr = get_string("plaintext: ");
    printf("ciphertext: ");

    int count = 0;                                              // count is the next key used
    for (int i = 0, plainlen = strlen(plainstr); i < plainlen; i++)
    {
        char c = plainstr[i];
        if (isalpha(c))                                         // if char is a letter,
        {
            if (isupper(c))                                         // check if it is capitalized
            {
                int cipher = ((((int) c - 65) + keyarr[count]) % 26) + 65;
                printf("%c", (char) cipher);
            }

            else                                                    // or not
            {
                int cipher = ((((int) c - 97) + keyarr[count]) % 26) + 97;
                printf("%c", (char) cipher);
            }

            count++;
            if (count == keylen)                                // reset "count" if no next key exists
            {
                count = 0;
            }
        }

        else                                        // if char is not a letter, print and do not increase "count"
        {
            printf("%c", plainstr[i]);
        }
    }

    printf("\n");
}

bool allalpha(string s)                 // check whether all chars in a string is alphabetical
{
    for (int i = 0, len = strlen(s); i < len; i++)
    {
        if (!isalpha(s[i]))
        {
            return false;
        }
    }
    return true;
}

File: vigenere.py
import sys

from cs50 import get_string

# wrong usage
if len(sys.argv) != 2 or not sys.argv[1].isalpha():
    print("Usage: python vigenere.py [key]")
    exit(1)

# make list of key
keylist = []
for c in sys.argv[1]:

    # uppercase char
    if c.isupper():
        keylist.append(ord(c) - 65)

    # lowercase char
    else:
        keylist.append(ord(c) - 97)

# prompt user for plain text
plain = get_string("plaintext: ")

# init cipher text and index
cipher = ""
index = 0

# loop through every char in plain text
for c in plain:

    # alphabetical char
    if c.isalpha():

        # uppercase letter
        if c.isupper():
            cipher += chr((ord(c) + keylist[index] - 65) % 26 + 65)

        # lowercase letter
        else:
            cipher += chr((ord(c) + keylist[index] - 97) % 26 + 97)

        # increase index, reset if it exceeds keylist length
        index += 1
        if index == len(keylist):
            index = 0

    # not alphabetical, keep the char
    else:
        cipher += c

# output the cipher text
print("ciphertext: " + cipher)
sys.exit(0)